examples/llm/src/mosaic_gpt.py

```diff
@@ -86,26 +87,74 @@ class FlashCausalAttention(nn.Module):
     def __init__(self, cfg: DictConfig, device: Optional[str] = None):
         super().__init__()
         try:
-            from flash_attn.flash_attention import FlashMHA  # type: ignore
+            from flash_attn.flash_attention import (  # type: ignore
+                FlashAttention, FlashMHA)
         except ImportError as e:
             raise e
 
-        self.mhsa = FlashMHA(
-            embed_dim=cfg.d_model,
-            num_heads=cfg.n_heads,
-            attention_dropout=cfg.attn_pdrop,
-            bias=True,
-            batch_first=True,
-            causal=True,
-            device=device,
-        )
-        self.mhsa.out_proj._is_residual = True
+        self.attn_qk_ln = cfg.attn_qk_ln
+        self.d_model = cfg.d_model
+        self.n_heads = cfg.n_heads
+
+        if self.attn_qk_ln:
+            self.W_qkv = nn.Linear(self.d_model,
+                                   3 * self.d_model,
+                                   bias=True,
+                                   device=device)
+            self.causal_attn = FlashAttention(attention_dropout=cfg.attn_pdrop,
+                                              device=device)
+            self.out_proj = nn.Linear(self.d_model,
+                                      self.d_model,
+                                      bias=True,
+                                      device=device)
+
+            self.out_proj._is_residual = True  # type: ignore
+
+            self.q_ln = nn.LayerNorm(self.d_model, device=device)
+            self.k_ln = nn.LayerNorm(self.d_model, device=device)
```

Comment:
Just want to double check here that 2 LayerNorms are required and not just a shared one for both activations.

Reply:
Yeah how it's formulated is separately